twilight-gateway/src/latency.rs

```diff
@@ -0,0 +1,182 @@
+//! Statistics about the latency of a shard, useful for debugging.
+
+use std::{
+    slice::Iter,
+    time::{Duration, Instant},
+};
+
+/// Information about the latency of a [`Shard`]'s websocket connection.
+///
+/// May be obtained via [`Shard::latency`].
+///
+/// [`Shard`]: crate::Shard
+/// [`Shard::latency`]: crate::Shard::latency
+#[derive(Clone, Debug)]
+pub struct Latency {
+    /// Total number of heartbeat acknowledgements that have been received
+    /// during the session.
+    heartbeats: u32,
+    /// When the last heartbeat acknowledgement was received.
+    received: Option<Instant>,
+    /// Most recent latencies between sending a heartbeat and receiving an
+    /// acknowledgement.
+    recent: [Duration; Self::RECENT_LEN],
+    /// When the last heartbeat was sent.
+    sent: Option<Instant>,
+    /// Combined latencies of all heartbeats, used in conjunction with
+    /// [`heartbeats`] to determine the average latency.
+    ///
+    /// [`heartbeats`]: Self::heartbeats
+    total_duration: Duration,
+}
+
+impl Latency {
+    /// Maximum number of recent latencies to store.
+    const RECENT_LEN: usize = 5;
+
+    /// Create a new instance for tracking shard latency.
+    pub(crate) const fn new() -> Self {
+        Self {
+            heartbeats: 0,
+            received: None,
+            recent: [Duration::ZERO; Self::RECENT_LEN],
+            sent: None,
+            total_duration: Duration::ZERO,
+        }
+    }
+
+    /// The average time it took to receive an acknowledgement for every
+    /// heartbeat sent over the duration of the session.
+    ///
+    /// For example, a reasonable value for this may be between 10 to 100
+    /// milliseconds depending on the network connection and physical location.
+    ///
+    /// # Note
+    ///
+    /// If this is None, the shard has not received a heartbeat yet.
+    pub const fn average(&self) -> Option<Duration> {
+        self.total_duration.checked_div(self.heartbeats)
+    }
+
+    /// The total number of heartbeats that have been sent during this session.
+    pub const fn heartbeats(&self) -> u32 {
+        self.heartbeats
+    }
+
+    /// The 5 most recent latency times.
+    ///
+    /// Index 0 is the oldest, 4 is the most recent.
+    pub fn recent(&self) -> Iter<'_, Duration> {
+        self.recent.iter()
+    }
```

Comment:
I don't see why this should not return an array reference (if we want `RECENT_LEN` to be part of the API) or slice. Especially with the current documentation referencing indexes.
```suggestion
    /// The 5 most recent latency times.
    ///
    /// Index 0 is the oldest, 4 is the most recent.
    pub fn recent(&self) -> &[Duration; Self::RECENT_LEN] {
        self.recent.iter()
    }
```